dumps/dtx_update_abort.c

```diff
@@ -28,36 +28,59 @@ vts_dtx_begin // a test-specific dtx_begin() altenative
                         /* dtx_is_valid_handle(dth) */
-                        vos_ts_set_allocate(&ioc->ic_ts_set, vos_flags, cflags, iod_nr, dth, cont->vc_pool->vp_sysdb);
+                        /* Allocate a timestamp set */
+                        vos_ts_set_allocate(ts_set = &ioc->ic_ts_set, vos_flags, cflags, iod_nr, dth, cont->vc_pool->vp_sysdb);
                                 /* dtx_is_valid_handle(dth) */
+                                /* dth->dth_local == false */
+                                tx_id = &dth->dth_xid;
+                                size = VOS_TS_TYPE_AKEY + akey_nr;
+	                        array_size = size * sizeof((*ts_set)->ts_entries[0]);
+                                D_ALLOC(*ts_set, sizeof(**ts_set) + array_size);
                                 /* tx_id != NULL */
                                 uuid_copy((*ts_set)->ts_tx_id.dti_uuid, tx_id->dti_uuid);
                                 (*ts_set)->ts_tx_id.dti_hlc = tx_id->dti_hlc;
                                 vos_ts_set_append_cflags(*ts_set, cflags);
                                         /* vos_ts_in_tx(ts_set) */
                         /* rc == 0 */
-                        bioc = vos_data_ioctxt(cont->vc_pool);
+                        bioc = vos_data_ioctxt(vp = cont->vc_pool);
                                 struct bio_meta_context *mc = vos_pool2mc(vp);
                                 /* mc == NULL */
-                        ioc->ic_biod = bio_iod_alloc(bioc, vos_ioc2umm(ioc), iod_nr, read_only ? BIO_IOD_TYPE_FETCH : BIO_IOD_TYPE_UPDATE);
+                                /* Use dummy I/O context when data blob doesn't exist */
+                                return vp->vp_dummy_ioctxt;
+                        ioc->ic_biod = bio_iod_alloc(bioc, vos_ioc2umm(ioc), sgl_cnt = iod_nr, read_only ? BIO_IOD_TYPE_FETCH : BIO_IOD_TYPE_UPDATE);
+                                D_ALLOC(biod, offsetof(struct bio_desc, bd_sgls[sgl_cnt]));
+                                return biod;
                         /* ioc->ic_biod != NULL */
-                        dcs_csum_info_list_init(&ioc->ic_csum_list, iod_nr);
+                        dcs_csum_info_list_init(list = &ioc->ic_csum_list, nr = iod_nr);
+                                daos_size_t initial_size = (sizeof(struct dcs_csum_info) + 8) * nr;
+                                list->dcl_buf_size = initial_size;
+                                memset(list, 0, sizeof(*list));
+                                D_ALLOC(list->dcl_csum_infos, list->dcl_buf_size);
                         for (i = 0; i < iod_nr; i++) {
-                                bsgl = bio_iod_sgl(ioc->ic_biod, i);
-                                bio_sgl_init(bsgl, iov_nr);
+                                bsgl = bio_iod_sgl(biod = ioc->ic_biod, i);
+                                        return &biod->bd_sgls[idx];
+                                bio_sgl_init(sgl = bsgl, nr = iov_nr);
```

Comment:
iod_nr?

Reply:
Done. Please find the definition I have added a few lines above.